File: frontend/src/lib/api/reports.ts
import { api, buildQuery } from '$lib/api/client';

export interface MilkSummary {
	total_milk: number;
	count_days: number;
	avg_per_animal: number | null;
}

export interface ReproductionSummary {
	total_calvings: number;
	total_inseminations: number;
	total_pregnancies: number;
	total_heats: number;
	total_dry_offs: number;
}

export interface FeedSummary {
	total_feed_kg: number;
	total_visits: number;
}

export interface HerdOverviewRow {
	date: string;
	cow_count: number;
	total_milk: number | null;
	avg_day_production: number | null;
	total_milkings: number | null;
	total_refusals: number | null;
	total_failures: number | null;
	milk_separated: number | null;
	avg_scc: number | null;
}

export interface HerdOverviewResponse {
	period: HerdOverviewRow[];
	avg_cow_count: number;
	avg_milk: number | null;
	avg_milkings: number | null;
	avg_failures: number | null;
	avg_scc: number | null;
}

export interface RestFeedRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	feed_date: string;
	feed_number: number;
	total_planned: number;
	rest_feed: number | null;
	rest_feed_pct: number | null;
}

export interface RestFeedResponse {
	rows: RestFeedRow[];
	total_rest_feed_pct: number | null;
}

export interface RobotPerformanceRow {
	device_address: number | null;
	date: string;
	avg_milk_speed: number | null;
	max_milk_speed: number | null;
	milkings: number;
	avg_lf_milk_time: number | null;
	avg_lr_milk_time: number | null;
	avg_rf_milk_time: number | null;
	avg_rr_milk_time: number | null;
	avg_lf_dead_milk_time: number | null;
	avg_lr_dead_milk_time: number | null;
	avg_rf_dead_milk_time: number | null;
	avg_rr_dead_milk_time: number | null;
}

export interface FailedMilkingRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	visit_datetime: string;
	device_address: number | null;
	milk_yield: number | null;
	lf_colour: string | null;
	lr_colour: string | null;
	rf_colour: string | null;
	rr_colour: string | null;
	lf_conductivity: number | null;
	lr_conductivity: number | null;
	rf_conductivity: number | null;
	rr_conductivity: number | null;
}

export interface UdderHealthRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	visit_datetime: string;
	lf_conductivity: number | null;
	lr_conductivity: number | null;
	rf_conductivity: number | null;
	rr_conductivity: number | null;
	lf_colour: string | null;
	lr_colour: string | null;
	rf_colour: string | null;
	rr_colour: string | null;
	latest_scc: number | null;
	milk_yield: number | null;
	deviation_day_prod: number | null;
	attention_quarters: string[];
	separation: string | null;
}

export interface MilkDayProductionTimeRow {
	date: string;
	total_milk: number | null;
	avg_milk_per_cow: number | null;
	cow_count: number;
	milkings: number | null;
	refusals: number | null;
	failures: number | null;
	avg_weight: number | null;
	total_feed: number | null;
	total_rest_feed: number | null;
}

export interface VisitBehaviorRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	total_milkings: number;
	total_refusals: number;
	avg_milk_per_milking: number | null;
	avg_duration_seconds: number | null;
	milk_frequency_setting: number | null;
	last_visit: string | null;
}

export interface CalendarCalvingRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	lac_number: number | null;
	group_number: number | null;
	last_insemination_date: string | null;
	expected_calving_date: string | null;
	days_until_calving: number | null;
	sire_code: string | null;
	days_pregnant: number | null;
}

export interface CalendarDryOffRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	expected_calving_date: string | null;
	recommended_dry_off_date: string | null;
	days_until_dry_off: number | null;
	lac_number: number | null;
}

export interface CalendarHeatRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	last_heat_date: string | null;
	expected_heat_date: string | null;
	days_until_heat: number | null;
	days_in_lactation: number | null;
	inseminated: boolean;
	overdue: boolean;
}

export interface CalendarPregnancyCheckRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	insemination_date: string | null;
	sire_code: string | null;
	days_since_insemination: number | null;
	pregnancy_confirmed: boolean;
}

export interface CalendarResponse {
	expected_calvings: CalendarCalvingRow[];
	expected_dry_offs: CalendarDryOffRow[];
	expected_heats: CalendarHeatRow[];
	pregnancy_checks: CalendarPregnancyCheckRow[];
}

export interface HealthActivityRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	health_index: number | null;
	activity_deviation: number | null;
	rumination_minutes: number | null;
	max_rumination_change_24h: number | null;
	rumination_3day_diff: number | null;
	latest_milk: number | null;
	avg_milk_7d: number | null;
	milk_deviation_pct: number | null;
}

export interface CowRobotEfficiencyRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	milk_per_box_time_week: number | null;
	avg_milk_speed: number | null;
	avg_treatment_time: number | null;
	avg_milking_time: number | null;
	milkings_7d: number;
	total_milk_7d: number | null;
	avg_milk_per_milking: number | null;
}

export interface LactationAnalysisPoint {
	dim: number;
	avg_milk: number | null;
	avg_visits: number | null;
	avg_feed: number | null;
	avg_weight: number | null;
	avg_fat: number | null;
	avg_protein: number | null;
	cow_count: number;
}

export interface LactationAnalysisResponse {
	lac_number: number;
	points: LactationAnalysisPoint[];
}

export interface FeedPerTypeDayRow {
	date: string;
	feed_type: string;
	feed_type_name: string;
	total_amount_product: number | null;
	total_amount_dm: number | null;
	total_cost: number | null;
	cost_per_100milk: number | null;
}

export interface FeedPerTypeResponse {
	rows: FeedPerTypeDayRow[];
	avg_cost_per_100milk: number | null;
	total_cost: number | null;
}

export interface FeedPerCowDayRow {
	date: string;
	animal_count: number;
	avg_total_per_cow: number | null;
	avg_concentrate_per_cow: number | null;
	avg_roughage_per_cow: number | null;
	avg_cost_per_cow: number | null;
	avg_rumination_minutes: number | null;
	avg_day_production: number | null;
	avg_lactation_days: number | null;
	feed_efficiency: number | null;
}

export interface HealthTaskRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	sick_chance: number;
	sick_chance_status: string;
	milk_drop_kg: number | null;
	conductivity_highest: number | null;
	conductivity_chronic_quarters: string[];
	scc_indication: number | null;
	activity_deviation: number | null;
	rumination_deviation: number | null;
	weight_trend: number | null;
	total_weight_loss: number | null;
	fat_protein_ratio: number | null;
	feed_rest_pct: number | null;
	temperature_highest: number | null;
	colour_attentions: string[];
	milk_trend_deviation: number | null;
	days_in_lactation: number | null;
}

export interface HealthTaskResponse {
	rows: HealthTaskRow[];
}

export interface PregnancyRatePeriod {
	end_date: string;
	eligible: number;
	inseminated: number;
	pregnant: number;
	insemination_rate: number | null;
	conception_rate: number | null;
	pregnancy_rate: number | null;
}

export interface PregnancyRateResponse {
	periods: PregnancyRatePeriod[];
}

export interface TransitionRow {
	animal_id: number;
	animal_name: string | null;
	life_number: string | null;
	days_relative: number;
	milk_24h: number | null;
	sick_chance: number | null;
	rumination_3day_diff: number | null;
	rumination_minutes: number | null;
	feed_total: number | null;
	feed_rest: number | null;
	latest_scc: number | null;
}

export interface TransitionResponse {
	rows: TransitionRow[];
}

export async function getMilkSummary(from?: string, till?: string, signal?: AbortSignal) {
	return api<MilkSummary>(
		`/reports/milk-summary${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getReproductionSummary(from?: string, till?: string, signal?: AbortSignal) {
	return api<ReproductionSummary>(
		`/reports/reproduction-summary${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getFeedSummary(from?: string, till?: string, signal?: AbortSignal) {
	return api<FeedSummary>(
		`/reports/feed-summary${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getHerdOverview(from?: string, till?: string, signal?: AbortSignal) {
	return api<HerdOverviewResponse>(
		`/reports/herd-overview${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getRestFeed(from?: string, till?: string, signal?: AbortSignal) {
	return api<RestFeedResponse>(
		`/reports/rest-feed${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getRobotPerformance(from?: string, till?: string, signal?: AbortSignal) {
	return api<RobotPerformanceRow[]>(
		`/reports/robot-performance${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getFailedMilkings(from?: string, till?: string, signal?: AbortSignal) {
	return api<FailedMilkingRow[]>(
		`/reports/failed-milkings${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getUdderHealthWorklist(signal?: AbortSignal) {
	return api<{ rows: UdderHealthRow[] }>('/reports/udder-health-worklist', { signal });
}

export async function getUdderHealthAnalyze(signal?: AbortSignal) {
	return api<{ rows: UdderHealthRow[] }>('/reports/udder-health-analyze', { signal });
}

export async function getMilkDayProductionTime(from?: string, till?: string, signal?: AbortSignal) {
	return api<MilkDayProductionTimeRow[]>(
		`/reports/milk-day-production-time${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getVisitBehavior(from?: string, till?: string, signal?: AbortSignal) {
	return api<VisitBehaviorRow[]>(
		`/reports/visit-behavior${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getCalendar(signal?: AbortSignal) {
	return api<CalendarResponse>('/reports/calendar', { signal });
}

export async function getHealthActivityRumination(signal?: AbortSignal) {
	return api<HealthActivityRow[]>('/reports/health-activity-rumination', { signal });
}

export async function getCowRobotEfficiency(signal?: AbortSignal) {
	return api<CowRobotEfficiencyRow[]>('/reports/cow-robot-efficiency', { signal });
}

export async function getLactationAnalysis(lacNumber?: number, signal?: AbortSignal) {
	return api<LactationAnalysisResponse[]>(
		`/reports/lactation-analysis${buildQuery({ lac_number: lacNumber })}`,
		{ signal },
	);
}

export async function getFeedPerTypeDay(from?: string, till?: string, signal?: AbortSignal) {
	return api<FeedPerTypeResponse>(
		`/reports/feed-per-type-day${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getFeedPerCowDay(from?: string, till?: string, signal?: AbortSignal) {
	return api<FeedPerCowDayRow[]>(
		`/reports/feed-per-cow-day${buildQuery({ from_date: from, till_date: till })}`,
		{ signal },
	);
}

export async function getHealthTask(signal?: AbortSignal) {
	return api<HealthTaskResponse>('/reports/health-task', { signal });
}

export async function getPregnancyRate(signal?: AbortSignal) {
	return api<PregnancyRateResponse>('/reports/pregnancy-rate', { signal });
}

export async function getTransitionReport(signal?: AbortSignal) {
	return api<TransitionResponse>('/reports/transition', { signal });
}

export function getExportUrl(
	type: 'milk' | 'reproduction' | 'feed',
	from?: string,
	till?: string,
	format?: 'csv' | 'pdf',
): string {
	const base = import.meta.env.VITE_API_BASE || '/api/v1';
	const suffix = format === 'pdf' ? `/${type}/pdf` : `/${type}`;
	return `${base}/reports/export${suffix}${buildQuery({ from_date: from, till_date: till })}`;
}

export function getReportExportUrl(
	reportType: string,
	from?: string,
	till?: string,
	format: 'csv' | 'pdf' = 'csv',
): string {
	const base = import.meta.env.VITE_API_BASE || '/api/v1';
	const ext = format === 'pdf' ? 'pdf' : 'csv';
	return `${base}/reports/export/${reportType}/${ext}${buildQuery({ from_date: from, till_date: till })}`;
}

File: backend/src/models/reports.rs
use serde::Serialize;

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct HerdOverviewRow {
    pub date: String,
    pub cow_count: i64,
    pub total_milk: Option<f64>,
    pub avg_day_production: Option<f64>,
    pub total_milkings: Option<i64>,
    pub total_refusals: Option<i64>,
    pub total_failures: Option<i64>,
    pub milk_separated: Option<i64>,
    pub avg_scc: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct HerdOverviewResponse {
    pub period: Vec<HerdOverviewRow>,
    pub avg_cow_count: f64,
    pub avg_milk: Option<f64>,
    pub avg_milkings: Option<f64>,
    pub avg_failures: Option<f64>,
    pub avg_scc: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct RestFeedRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub feed_date: String,
    pub feed_number: i32,
    pub total_planned: f64,
    pub rest_feed: Option<i32>,
    pub rest_feed_pct: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct RestFeedResponse {
    pub rows: Vec<RestFeedRow>,
    pub total_rest_feed_pct: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct CowDailyProductionRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub date: String,
    pub milk_amount: Option<f64>,
    pub avg_amount: Option<f64>,
    pub avg_weight: Option<f64>,
    pub isk: Option<f64>,
    pub scc: Option<i32>,
    pub fat_pct: Option<f64>,
    pub protein_pct: Option<f64>,
    pub lactose_pct: Option<f64>,
    pub feed_total: Option<f64>,
    pub feed_rest: Option<i32>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct RobotPerformanceRow {
    pub device_address: Option<i32>,
    pub date: String,
    pub avg_milk_speed: Option<f64>,
    pub max_milk_speed: Option<f64>,
    pub milkings: i64,
    pub avg_lf_milk_time: Option<f64>,
    pub avg_lr_milk_time: Option<f64>,
    pub avg_rf_milk_time: Option<f64>,
    pub avg_rr_milk_time: Option<f64>,
    pub avg_lf_dead_milk_time: Option<f64>,
    pub avg_lr_dead_milk_time: Option<f64>,
    pub avg_rf_dead_milk_time: Option<f64>,
    pub avg_rr_dead_milk_time: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct FailedMilkingRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub visit_datetime: String,
    pub device_address: Option<i32>,
    pub milk_yield: Option<f64>,
    pub lf_colour: Option<String>,
    pub lr_colour: Option<String>,
    pub rf_colour: Option<String>,
    pub rr_colour: Option<String>,
    pub lf_conductivity: Option<i32>,
    pub lr_conductivity: Option<i32>,
    pub rf_conductivity: Option<i32>,
    pub rr_conductivity: Option<i32>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct UdderHealthRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub visit_datetime: String,
    pub lf_conductivity: Option<i32>,
    pub lr_conductivity: Option<i32>,
    pub rf_conductivity: Option<i32>,
    pub rr_conductivity: Option<i32>,
    pub lf_colour: Option<String>,
    pub lr_colour: Option<String>,
    pub rf_colour: Option<String>,
    pub rr_colour: Option<String>,
    pub latest_scc: Option<i32>,
    pub milk_yield: Option<f64>,
    pub deviation_day_prod: Option<f64>,
    pub attention_quarters: Vec<String>,
    pub separation: Option<String>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct UdderHealthResponse {
    pub rows: Vec<UdderHealthRow>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct MilkDayProductionTimeRow {
    pub date: String,
    pub total_milk: Option<f64>,
    pub avg_milk_per_cow: Option<f64>,
    pub cow_count: i64,
    pub milkings: Option<i64>,
    pub refusals: Option<i64>,
    pub failures: Option<i64>,
    pub avg_weight: Option<f64>,
    pub total_feed: Option<f64>,
    pub total_rest_feed: Option<i64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct VisitBehaviorRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub total_milkings: i64,
    pub total_refusals: i64,
    pub avg_milk_per_milking: Option<f64>,
    pub avg_duration_seconds: Option<f64>,
    pub milk_frequency_setting: Option<i32>,
    pub last_visit: Option<String>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct CalendarCalvingRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub lac_number: Option<i32>,
    pub group_number: Option<i32>,
    pub last_insemination_date: Option<String>,
    pub expected_calving_date: Option<String>,
    pub days_until_calving: Option<i64>,
    pub sire_code: Option<String>,
    pub days_pregnant: Option<i64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct CalendarDryOffRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub expected_calving_date: Option<String>,
    pub recommended_dry_off_date: Option<String>,
    pub days_until_dry_off: Option<i64>,
    pub lac_number: Option<i32>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct CalendarHeatRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub last_heat_date: Option<String>,
    pub expected_heat_date: Option<String>,
    pub days_until_heat: Option<i64>,
    pub days_in_lactation: Option<i64>,
    pub inseminated: bool,
    pub overdue: bool,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct CalendarPregnancyCheckRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub insemination_date: Option<String>,
    pub sire_code: Option<String>,
    pub days_since_insemination: Option<i64>,
    pub pregnancy_confirmed: bool,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct CalendarResponse {
    pub expected_calvings: Vec<CalendarCalvingRow>,
    pub expected_dry_offs: Vec<CalendarDryOffRow>,
    pub expected_heats: Vec<CalendarHeatRow>,
    pub pregnancy_checks: Vec<CalendarPregnancyCheckRow>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct HealthActivityRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub health_index: Option<f64>,
    pub activity_deviation: Option<f64>,
    pub rumination_minutes: Option<i32>,
    pub max_rumination_change_24h: Option<i32>,
    pub rumination_3day_diff: Option<i32>,
    pub latest_milk: Option<f64>,
    pub avg_milk_7d: Option<f64>,
    pub milk_deviation_pct: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct CowRobotEfficiencyRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub milk_per_box_time_week: Option<f64>,
    pub avg_milk_speed: Option<f64>,
    pub avg_treatment_time: Option<f64>,
    pub avg_milking_time: Option<f64>,
    pub milkings_7d: i64,
    pub total_milk_7d: Option<f64>,
    pub avg_milk_per_milking: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct LactationAnalysisPoint {
    pub dim: i32,
    pub avg_milk: Option<f64>,
    pub avg_visits: Option<f64>,
    pub avg_feed: Option<f64>,
    pub avg_weight: Option<f64>,
    pub avg_fat: Option<f64>,
    pub avg_protein: Option<f64>,
    pub cow_count: i64,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct LactationAnalysisResponse {
    pub lac_number: i32,
    pub points: Vec<LactationAnalysisPoint>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct FeedPerTypeDayRow {
    pub date: String,
    pub feed_type: String,
    pub feed_type_name: String,
    pub total_amount_product: Option<f64>,
    pub total_amount_dm: Option<f64>,
    pub total_cost: Option<f64>,
    pub cost_per_100milk: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct FeedPerTypeResponse {
    pub rows: Vec<FeedPerTypeDayRow>,
    pub avg_cost_per_100milk: Option<f64>,
    pub total_cost: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct FeedPerCowDayRow {
    pub date: String,
    pub animal_count: i64,
    pub avg_total_per_cow: Option<f64>,
    pub avg_concentrate_per_cow: Option<f64>,
    pub avg_roughage_per_cow: Option<f64>,
    pub avg_cost_per_cow: Option<f64>,
    pub avg_rumination_minutes: Option<f64>,
    pub avg_day_production: Option<f64>,
    pub avg_lactation_days: Option<f64>,
    pub feed_efficiency: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct HealthTaskRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub sick_chance: f64,
    pub sick_chance_status: String,
    pub milk_drop_kg: Option<f64>,
    pub conductivity_highest: Option<i32>,
    pub conductivity_chronic_quarters: Vec<String>,
    pub scc_indication: Option<i32>,
    pub activity_deviation: Option<f64>,
    pub rumination_deviation: Option<i32>,
    pub weight_trend: Option<f64>,
    pub total_weight_loss: Option<f64>,
    pub fat_protein_ratio: Option<f64>,
    pub feed_rest_pct: Option<f64>,
    pub temperature_highest: Option<f64>,
    pub colour_attentions: Vec<String>,
    pub milk_trend_deviation: Option<f64>,
    pub days_in_lactation: Option<i64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct HealthTaskResponse {
    pub rows: Vec<HealthTaskRow>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct PregnancyRatePeriod {
    pub end_date: String,
    pub eligible: i64,
    pub inseminated: i64,
    pub pregnant: i64,
    pub insemination_rate: Option<f64>,
    pub conception_rate: Option<f64>,
    pub pregnancy_rate: Option<f64>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct PregnancyRateResponse {
    pub periods: Vec<PregnancyRatePeriod>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct TransitionRow {
    pub animal_id: i32,
    pub animal_name: Option<String>,
    pub life_number: Option<String>,
    pub days_relative: i64,
    pub milk_24h: Option<f64>,
    pub sick_chance: Option<f64>,
    pub rumination_3day_diff: Option<i32>,
    pub rumination_minutes: Option<i32>,
    pub feed_total: Option<f64>,
    pub feed_rest: Option<i32>,
    pub latest_scc: Option<i32>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct TransitionResponse {
    pub rows: Vec<TransitionRow>,
}
